// @skpm/builder/lib/utils/babelLoader.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function babelLoader(userDefinedBabelConfig) {
  return {
    test: /\.jsx?$/,
    exclude: /node_modules/,
    use: {
      loader: 'babel-loader',
      options: _extends({
        babelrc: false,
        presets: [[require('@skpm/babel-preset')]]
      }, userDefinedBabelConfig || {})
    }
  };
}

exports.default = skpmConfig => {
  const babelrcPath = _path2.default.join(process.cwd(), '.babelrc');
  let userDefinedBabelConfig = null;
  try {
    if (_fs2.default.existsSync(babelrcPath)) {
      userDefinedBabelConfig = JSON.parse(_fs2.default.readFileSync(babelrcPath, 'utf8'));
    } else if (skpmConfig.babel) {
      userDefinedBabelConfig = skpmConfig.babel;
    }
  } catch (err) {
    console.error(`${_chalk2.default.red('error')} Error while reading babelrc`);
    console.error(err);
    process.exit(1);
  }

  return babelLoader(userDefinedBabelConfig);
};
