// @skpm/builder/lib/utils/webpackConfig.js
'use strict';

exports.__esModule = true;

let getCommands = (() => {
  var _ref = _asyncToGenerator(function* (output, commandIdentifiers) {
    return Promise.all(commandIdentifiers.map(function (commandIdentifier) {
      return (0, _webpackCommandPlugin2.default)(output, commandIdentifier);
    }));
  });

  return function getCommands(_x, _x2) {
    return _ref.apply(this, arguments);
  };
})();

// avoid looking it up every time


exports.default = getWebpackConfig;

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackMerge = require('webpack-merge');

var _webpackMerge2 = _interopRequireDefault(_webpackMerge);

var _uglifyjsWebpackPlugin = require('uglifyjs-webpack-plugin');

var _uglifyjsWebpackPlugin2 = _interopRequireDefault(_uglifyjsWebpackPlugin);

var _webpackCommandPlugin = require('./webpackCommandPlugin');

var _webpackCommandPlugin2 = _interopRequireDefault(_webpackCommandPlugin);

var _webpackHeaderFooterPlugin = require('./webpackHeaderFooterPlugin');

var _webpackHeaderFooterPlugin2 = _interopRequireDefault(_webpackHeaderFooterPlugin);

var _babelLoader = require('./babelLoader');

var _babelLoader2 = _interopRequireDefault(_babelLoader);

var _resourceLoader = require('./resourceLoader');

var _resourceLoader2 = _interopRequireDefault(_resourceLoader);

var _nibLoader = require('./nibLoader');

var _nibLoader2 = _interopRequireDefault(_nibLoader);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const CORE_MODULES = ['util', 'events', 'console'];

const isProd = process.env.NODE_ENV === 'production';

function getWebpackConfig(argv, output, manifestFolder, skpmConfig) {
  let userDefinedWebpackConfig;
  try {
    if (_fs2.default.existsSync(_path2.default.join(process.cwd(), 'webpack.skpm.config.js'))) {
      userDefinedWebpackConfig = require(_path2.default.join(process.cwd(), 'webpack.skpm.config.js'));
    }
  } catch (err) {
    console.error(`${_chalk2.default.red('error')} Error while reading webpack.skpm.config.js`);
    console.error(err);
    process.exit(1);
  }

  const babelLoader = (0, _babelLoader2.default)(skpmConfig);

  return (() => {
    var _ref2 = _asyncToGenerator(function* (file, commandIdentifiers, commandHandlers) {
      const basename = _path2.default.basename(file);

      let plugins = [new _webpack2.default.EnvironmentPlugin({
        NODE_ENV: 'development' // default to 'development'
      })];
      const rules = [babelLoader];

      if (commandIdentifiers) {
        if (commandHandlers.find(function (k) {
          return k === '__skpm_run';
        })) {
          console.error(`${_chalk2.default.red('error')} __skpm_run is a reserved keyword for handlers. Please use something else.`);
          process.exit(1);
        }

        plugins.push(new _webpack2.default.ProvidePlugin({
          fetch: require.resolve('sketch-polyfill-fetch'),
          setTimeout: [require.resolve('@skpm/timers/timeout'), 'setTimeout'],
          clearTimeout: [require.resolve('@skpm/timers/timeout'), 'clearTimeout'],
          setImmediate: [require.resolve('@skpm/timers/immediate'), 'setImmediate'],
          clearImmediate: [require.resolve('@skpm/timers/immediate'), 'clearImmediate'],
          setInterval: [require.resolve('@skpm/timers/interval'), 'setInterval'],
          clearInterval: [require.resolve('@skpm/timers/interval'), 'clearInterval'],
          Promise: require.resolve('promise-polyfill')
        }), new _webpackHeaderFooterPlugin2.default(commandHandlers));

        rules.push(_resourceLoader2.default);
        rules.push(_nibLoader2.default);
      }

      if (argv.run && commandIdentifiers) {
        plugins = plugins.concat((yield getCommands(output, commandIdentifiers)));
      }

      if (isProd) {
        plugins.push(new _uglifyjsWebpackPlugin2.default({
          uglifyOptions: {
            mangle: {
              // @see https://bugs.webkit.org/show_bug.cgi?id=171041
              // @see https://github.com/mishoo/UglifyJS2/issues/1753#issuecomment-324814782
              safari10: true
            }
          }
        }));
      }

      let webpackConfig = {
        mode: 'development',
        devtool: isProd ? 'none' : 'source-map',
        module: {
          rules
        },
        resolve: {
          extensions: ['.sketch.js', '.js'],
          modules: ['node_modules', _path2.default.join(__dirname, '..', '..', 'node_modules')]
        },
        resolveLoader: {
          modules: [_path2.default.join(__dirname, '..', '..', 'node_modules'), 'node_modules']
        },
        entry: _path2.default.join(commandIdentifiers ? manifestFolder : process.cwd(), file),
        externals: [function (context, request, callback) {
          // sketch API
          if (/^sketch\//.test(request) || request === 'sketch') {
            return callback(null, `commonjs ${request}`);
          }
          // core modules shipped in Sketch
          if (CORE_MODULES.indexOf(request) !== -1) {
            return callback(null, `commonjs ${request}`);
          }
          return callback();
        }],
        output: {
          filename: basename,
          library: commandIdentifiers ? 'exports' : undefined,
          path: commandIdentifiers ? _path2.default.join(output, 'Contents', 'Sketch') : _path2.default.join(output, 'Contents', 'Resources')
        },
        plugins
      };

      if (userDefinedWebpackConfig) {
        const resolvedUserDefinedConfig = yield userDefinedWebpackConfig(webpackConfig, !!commandIdentifiers);
        if (resolvedUserDefinedConfig) {
          webpackConfig = _webpackMerge2.default.smart(webpackConfig, resolvedUserDefinedConfig);
        }
      }

      return webpackConfig;
    });

    function webpackConfigGenerator(_x3, _x4, _x5) {
      return _ref2.apply(this, arguments);
    }

    return webpackConfigGenerator;
  })();
}
